util/pingpong.c

```diff
@@ -0,0 +1,2111 @@
+#define FT_WARN(fmt, ...) FT_LOG("warn", fmt, ##__VA_ARGS__)
+
+int _activate_debug = 0;
+#define FT_ACTIVATE_DEBUG _activate_debug
+
+#define FT_DEBUG(fmt, ...)						\
+	do {								\
+		if (_activate_debug) {					\
+			FT_LOG_("debug", fmt, ##__VA_ARGS__);		\
+		}							\
+	} while (0)
+
+#define FT_EQ_ERR(eq, entry, buf, len) \
+	FT_ERR("eq_readerr: %s", fi_eq_strerror(eq, entry.prov_errno, \
+				entry.err_data, buf, len))
+
+#define FT_CQ_ERR(cq, entry, buf, len) \
+	FT_ERR("cq_readerr: %s", fi_cq_strerror(cq, entry.prov_errno, \
+				entry.err_data, buf, len))
+
+#define FT_CLOSE_FID(fd)						\
+	do {								\
+		int ret;						\
+		if ((fd)) {						\
+			ret = fi_close(&(fd)->fid);			\
+			if (ret)					\
+				FT_ERR("fi_close (%d) fid %d",		\
+					ret, (int) (fd)->fid.fclass);	\
+			fd = NULL;					\
+		}							\
+	} while (0)
+
+#define FT_CLOSEV_FID(fd, cnt)			\
+	do {					\
+		int i;				\
+		if (!(fd))			\
+			break;			\
+		for (i = 0; i < (cnt); i++) {	\
+			FT_CLOSE_FID((fd)[i]);	\
+		}				\
+	} while (0)
+
+
+#define FT_PROCESS_QUEUE_ERR(readerr, rd, queue, fn, str)	\
+	do {							\
+		if (rd == -FI_EAVAIL) {				\
+			readerr(queue, fn " " str);		\
+		} else {					\
+			FT_PRINTERR(fn, rd);			\
+		}						\
+	} while (0)
+
+#define FT_PROCESS_EQ_ERR(rd, eq, fn, str) \
+	FT_PROCESS_QUEUE_ERR(eq_readerr, rd, eq, fn, str)
+
+#define FT_PRINT_OPTS_USAGE(opt, desc) fprintf(stderr, " %-20s %s\n", opt, desc)
+
+#define MIN(a,b) (((a)<(b))?(a):(b))
+#define MAX(a,b) (((a)>(b))?(a):(b))
+#define ARRAY_SIZE(A) (sizeof(A)/sizeof(*A))
+
+#define TEST_ENUM_SET_N_RETURN(str, enum_val, type, data)	\
+	TEST_SET_N_RETURN(str, #enum_val, enum_val, type, data)
+
+#define TEST_SET_N_RETURN(str, val_str, val, type, data)	\
+	do {							\
+		if (!strncmp(str, val_str, strlen(val_str))) {	\
+			*(type *)(data) = val;			\
+			return 0;				\
+		}						\
+	} while (0)
+
+
+#define CTRL_BUF_LEN 64
+
+struct ct_pingpong {
+	struct fi_info *fi_pep, *fi, *hints;
+	struct fid_fabric *fabric;
+	struct fid_domain *domain;
+	struct fid_pep *pep;
+	struct fid_ep *ep;
+	struct fid_cq *txcq, *rxcq;
+	struct fid_mr *mr;
+	struct fid_av *av;
+	struct fid_eq *eq;
+
+	struct fid_mr no_mr;
+	struct fi_context tx_ctx, rx_ctx;
+	struct fi_context *ctx_arr;
+	uint64_t remote_cq_data;
+
+	uint64_t tx_seq, rx_seq, tx_cq_cntr, rx_cq_cntr;
+	//int ft_skip_mr;
+	int ft_parent_proc;
+	pid_t ft_child_pid;
+
+	fi_addr_t remote_fi_addr;
+	void *buf, *tx_buf, *rx_buf;
+	size_t buf_size, tx_size, rx_size;
+	int rx_fd, tx_fd;
+	int data_default_port;
+	char data_port[8];
+
+	char test_name[50];
+	int timeout;
+	struct timespec start, end;
+
+	struct fi_av_attr av_attr;
+	struct fi_eq_attr eq_attr;
+	struct fi_cq_attr cq_attr;
+	struct ft_opts opts;
+
+	long cnt_ack_msg;
+
+	int ctrl_port;
+	int ctrl_listenfd;
+	int ctrl_connfd;
+	struct sockaddr_in ctrl_addr;
+	char ctrl_buf[CTRL_BUF_LEN + 1];
+};
+
+#define FT_ENABLE_ALL		(~0)
+#define FT_DEFAULT_SIZE		(1 << 0)
```

Comment:
Indeed, most of the universe is comprised of void in spite of our disappointment.  Some of this space must fill some part of out source code too.
Then the only question is : Do we accept the void even if it outlines and gives structure to the non-void ?
Or is it the other way around ?  :stuck_out_tongue_winking_eye: